gpt_engineer/ai.py

```diff
@@ -50,8 +65,64 @@ def next(self, messages: list[dict[str, str]], prompt=None):
         print()
         messages += [{"role": "assistant", "content": "".join(chat)}]
         logger.debug(f"Chat completion finished: {messages}")
+
+        self.update_token_usage_log(messages=messages, answer="".join(chat),
+                                    step_name=step_name)
+
         return messages
 
+    def update_token_usage_log(self, messages, answer, step_name):
+        prompt_tokens = self.num_tokens_from_messages(messages)
+        completion_tokens = self.num_tokens(answer)
+        total_tokens = prompt_tokens + completion_tokens
+
+        self.cumulative_prompt_tokens += prompt_tokens
+        self.cumulative_completion_tokens += completion_tokens
+        self.cumulative_total_tokens += total_tokens
+
+        self.token_usage_log.append({
+            "step_name": step_name,
```

Comment:
Would be great to use a `dataclass` here to make it easier to get IDE support etc!

Reply:
agree - done ✅